ext/mbstring/libmbfl/filters/mbfilter_gb18030.c

```diff
@@ -437,32 +458,39 @@ static size_t mb_gb18030_to_wchar(unsigned char **in, size_t *in_len, uint32_t *
 			} else if (c >= 0xA1 && c <= 0xA7 && c2 >= 0x40 && c2 < 0xA1 && c2 != 0x7F) {
 				/* UDA part 3: U+E4C6-U+E765 */
 				*out++ = 96*(c - 0xA1) + c2 - (c2 >= 0x80 ? 0x41 : 0x40) + 0xE4C6;
-			} else {
-				if ((s >= 0xA2AB && s <= 0xA9FE) || (s >= 0xD7FA && s <= 0xD7FE) || (s >= 0xFE50 && s <= 0xFEA0)) {
-					for (int i = 0; i < mbfl_gb18030_pua_tbl_max; i++) {
-						if (s >= mbfl_gb18030_pua_tbl[i][2] && s <= mbfl_gb18030_pua_tbl[i][2] + mbfl_gb18030_pua_tbl[i][1] - mbfl_gb18030_pua_tbl[i][0]) {
-							*out++ = s - mbfl_gb18030_pua_tbl[i][2] + mbfl_gb18030_pua_tbl[i][0];
-							goto next_iteration;
+			} else if (c2 >= 0x40 && c2 != 0x7F && c2 != 0xFF) {
+				unsigned int w = (c - 0x81)*192 + c2 - 0x40;
+
+				if (w >= 0x192B) {
+					if (w <= 0x1EBE) {
+						if (w != 0x1963 && w != 0x1DBF && (w < 0x1E49 || w > 0x1E55) && w != 0x1E7F) {
+							*out++ = cp936_pua_tbl1[w - 0x192B];
+							continue;
+						}
+					} else if (w >= 0x413A) {
+						if (w <= 0x413E) {
+							*out++ = cp936_pua_tbl2[w - 0x413A];
+							continue;
+						} else if (w >= 0x5DD0 && w <= 0x5E20) {
+							unsigned int c = gb18030_pua_tbl3[w - 0x5DD0];
+							if (c) {
+								*out++ = c;
+								continue;
+							}
 						}
```

Comment:
I was talking about this whole bit being black magic, but let me see if I understand. You are mapping certain bytes to a code point to different lookup tables by shitting the byte value?

Reply:
It's like this...

Specific ranges of byte values are used for 2-byte GB18030 characters. In this branch of the code, we're looking at values from `0x8140-817E`, `0x8180-81FE`, `0x8240-827E`, `0x8280-82FE`, `0x8340-837E`... and so on.

So first thing, convert those 2 byte values to a contiguous index into a lookup table.

But then, we really want different lookup tables for CP936/GB18030 and for HZ/EUC-CN, because for CP936/GB18030 we use PUA codepoints for certain characters, which we don't do for HZ/EUC-CN.

But because the lookup table is big, and only 3 portions of it have some different values, we "mask off" those 3 portions of the lookup table by testing if the lookup index falls within certain ranges, and if it does, do the lookup in a different, smaller table instead.

Reply:
One thing... I haven't looked at the disassembly of this code to make sure, but code like this:

```
cp936_pua_tbl1[w - 0x192B];
```

...should be very efficient, because the compiler can convert that to `(cp936_pua_tbl1 - 0x192B)[w]` instead.

It happened to me before that I tried to do that optimization manually, and found the code ran *exactly* as fast as it did before. :smile:

Reply:
Okay thank you for the explanation :)